src/MySqlConnector/Core/ResultSet.cs

```diff
@@ -116,42 +116,55 @@ public async Task ReadResultSetHeaderAsync(IOBehavior ioBehavior)
 				}
 				else
 				{
-					static int ReadColumnCount(ReadOnlySpan<byte> span)
+					var columnCountPacket = ColumnCountPayload.Create(payload.Span, Session.SupportsMetaSkip);
+					if (!columnCountPacket.MetadataFollows)
 					{
-						var reader = new ByteArrayReader(span);
-						var columnCount_ = (int) reader.ReadLengthEncodedInteger();
-						if (reader.BytesRemaining != 0)
-							throw new MySqlException("Unexpected data at end of column_count packet; see https://github.com/mysql-net/MySqlConnector/issues/324");
-						return columnCount_;
+						// reuse previous metadata
+						var preparedStatement = DataReader.CurrentPrepareStmt()!;
+						ColumnDefinitions = preparedStatement.Columns!;
+						ColumnTypes = new MySqlDbType[columnCountPacket.ColumnCount];
+						for (var column = 0; column < columnCountPacket.ColumnCount; column++)
+						{
+							ColumnTypes[column] = TypeMapper.ConvertToMySqlDbType(ColumnDefinitions[column],
+								Connection.TreatTinyAsBoolean, Connection.GuidFormat);
+						}
```

Comment:
Note to self: Can we move this outside the `if` statement (and remove the corresponding code from the `else` branch)?

Reply:
I think best would be to get rid of the ColumnTypes variable, saving some calculation that are useful for metadata only, so that may not be used at all. => save ConnectionSettings and just rely on ColumnDefinitions

Reply:
Agreed: https://github.com/mysql-net/MySqlConnector/commit/5f72a2f741629fc5c2fa7781621315a329eb941d.